### ISSUE TITLE
query: rework as attrs-based model

### ISSUE THREAD

**Issue by synrg** [dronefly-garden/dronefly-core#1]
After studying pyinaturalist's models, I'd like Query to be based on pyinaturalist.base.models, as they provide robust abstractions that improve on what we've made so far. While pyinat does not have a user's query as a concept distinct from the API requests that would be needed to fulfill them, it does have core classes that we can use to make one.

A Query is not, itself, a description of a single API request. It is a text description of a number of parameters sent to iNaturalist to produce a single display with two parts, the *description of the base entity* of the request, and zero or more individual results relating to that entity (e.g. total # of observations for the query, and per-observer counts of observations and species for the query).

For example, a `fungi by me from ns in prj lichens atlantic on today` request could be realized as pyinat RequestParams with individual params as follows (slightly simplified for illustration purposes):

- `taxon = Taxon(id=47170)`
  - the `taxon` here is the first record returned from `/v1/taxa/autocomplete?q=fungi`
- `user = User(id=545640)`
  - the special keyword `me` refers to user's own id
- `place = Place(id=6853)`
  - `ns` is looked up in a table associated with the user's command context
- `project = Project(id=62291)`
  - `lichens atlantic` is the first matching record from `/v1/projects/autocomplete?q=lichens+atlantic`
- `observed_on = datetime.today()`
  - `today` is parsed via `dateparser.parse()`

For each of the entities retrieved from local tables, only a partial object is needed, just so that `place.id`, etc. will work.

In the `dronefly` codebase, this "fully parsed" query is called a `QueryResponse` (which I'm not entirely happy with). It is still, however, only a template for one or more primary requests for the page to fill it with content.

Which requests are performed depend on what command handles the query. For example, in this simplified rendering of a `taxon` display with the above query arguments:

> Kingdom Fungi (Fungi Including Lichens)
> is a kingdom with 52 observations in Lichens of Atlantic Canada from Nova Scotia, CA observed on Jan 2, 2023.
> __obs# (spp#) by user:__
> 3 (2) benarmstrong

Several distinct API *requests* based on the query would be needed to fill in all the parts above including at least:
- `/v1/observations?taxon_id=47170&user_id=545740&place_id=6853&project_id=62291&verifiable=any&observed_on=2023-01-02&per_page=0`
  - to get `total_results` to put in `52 observations`
- `/v1/observations?taxon_id=47170&user_id=545740&place_id=6853&project_id=62291&verifiable=any&observed_on=2023-01-02&user_id=545640&per_page=0`
  AND
  `/v1/observations/species_counts?taxon_id=47170&user_id=545740&place_id=6853&project_id=62291&verifiable=any&observed_on=2023-01-02&user_id=545640&per_page=0`
  - to get "3 (2)"
- with luck, the User record for `benarmstrong` is already cached, otherwise a `/v1/users/545740` might be needed to obtain this from `user.login`
- not to mention also the Place and Project API *requests* to obtain their names as well (though they, too, might already be cached)

Finally, it should be possible to map between a command with query argument to a URL to the web page that best represents that base request, and any other parts of the page (usually counts of each entity which link to searches for those entities on the web):
- https://www.inaturalist.org/taxa/47170 (the title of the display, "Kingdom Fungi")
- https://www.inaturalist.org/observations?verifiable=any&taxon_id=47170&project_id=15702&place_id=6853&on=2023-01-02 (the `52 observations` link)
- https://www.inaturalist.org/observations?verifiable=any&taxon_id=47170&user_id=545640&project_id=15702&place_id=6853&on=2023-01-02 (the `3 (2) benarmstrong` link)

With all this in mind, the Query class should represent all of these arguments in a way that more closely resembles existing pyinaturalist models.

Here is a representation of this progression from text to parsed query to a validated query that is finally ready to be used in a command as dict-like results from each step:

```py
>>> query = Query.parse("fungi by me from ns in prj lichens atlantic on today")
{
  "taxon": "birds",
  "user": "me",
  "place": "ns",
  "project": "lichens atlantic",
  "observed_on": "today",
}
>>> validated_query = query.validate()
{
  "taxon": { id: 3 },
  "user": { id: 545640 },
  "place": { id: 6853 },
  "project": { id: 15702 },
  "observed_on": "2023-01-02",
}
```

I'm still not sure of QueryResponse vs. some better name. Maybe ValidatedQuery?

**Comment by synrg** [comment]
I think "validation" is the wrong word. Really, this is just stages of evaluation in a sort of assembly-line fashion. The Query itself is mostly just an expression of a set of filters to apply to the result set and not fully realizable until used in a command which provides the "verb", as well as determining which presentation is appropriate for the results e.g. "search" (implictly observations) to produce a result set containing all matching observations, in a paginated display presentation.

While fully developed classes for these concepts aren't yet firm, this is a rough outline of the parts of a command as they stand today:

- A Dronefly Command is a request for a result set containing the entity or entities asked for by the Command (verb) and qualified by one or more preposition+noun phrases that make up the Query.
- Those phrases are mostly filters to narrow down which subset of entities are shown, but also may contain presentation qualifiers, such as "sort by obs" and "desc" that alter the sort order.
- The QueryResponse is an intermediate object that contains one or more preposition + noun phrases (modelled as option + argument pairs). At this point, everything in the Command string that needed to be looked up has been looked up. Everything awaits further processing to produce the final result set.
- We might call such a thing more accurately a QueryResults, except it is not yet the full results that the user asked for. That will only finally be obtained by applying the Command verb to it.

That might not be such a bad name after all, QueryResults, until it is made concrete via using it in a Command, so:

```py
query = Query.parse('fungi by me from ns in prj lichens atlantic on today")
taxon_command = Command("taxon")
counts_command = Command("taxon counts")
query_results = query.prepare() # => QueryResults
taxon = taxon_command.realize(query_results) # => Taxon
taxon_counts = counts_command.realize(query_results) # => TaxonCounts
menu = TaxonMenuWithCounts(taxon_results=taxon_result, taxon_counts_list=[taxon_counts])
menu.start()
```

**Comment by synrg** [comment]
Alternatively, I might prefer if we called it a PreparedQuery which shifts emphasis off of the "results" aspect of it (i.e. these aren't yet the final result, but an intermediate stage to fetching them), and thus:

```py
query = Query.parse('fungi by me from ns in prj lichens atlantic on today")
taxon_command = Command("taxon")
counts_command = Command("taxon counts")
prepared_query = query.prepare() # => PreparedQuery
taxon = taxon_command.realize(prepared_query) # => Taxon
user_or_place_taxon_counts = counts_command.realize(prepared_query) # => Union[UserTaxonCounts, PlaceTaxonCounts]
menu = TaxonMenuWithCounts(taxon=taxon, counts=[user_or_place_taxon_counts])
menu.start()
```

**Comment by synrg** [comment]
Or maybe I have it ass-backwards? We don't realize (or finalize) the command. We realize / finalize the prepared_query! OK, one more take ...

```py
query = Query.parse('fungi by me from ns in prj lichens atlantic on today")
taxon_command = Command("taxon")
counts_command = Command("taxon counts")
prepared_query = query.prepare() # => PreparedQuery
taxon = prepared_query.(prepared_query) # => Taxon
user_or_place_taxon_counts = counts_command.realize(prepared_query) # => Union[UserTaxonCounts, PlaceTaxonCounts]
menu = TaxonMenuWithCounts(taxon=taxon, counts=[user_or_place_taxon_counts])
menu.start()
```

**Comment by synrg** [comment]
Oops, misclick. Still not happy with this. Will return to it later. "realizing" a command seems wrong. perhaps the two inputs to the menu are just results of functions applied to the prepared_query instead.

**Comment by synrg** [comment]
My next take relies far fewer new Dronefly objects and instead directly makes use of existing pyinat [TaxonCounts](https://pyinaturalist.readthedocs.io/en/latest/modules/pyinaturalist.models.TaxonCounts.html#taxoncounts) and [TaxonCount](https://pyinaturalist.readthedocs.io/en/stable/modules/pyinaturalist.models.TaxonCount.html#pyinaturalist.models.TaxonCount) models. 

We're back to a single command instead of a primary `taxon` command coupled with a secondary `taxon counts` command afterwards. The `taxon` command proceeds, bucket-brigade style, through three stages. It starts with parsing the query & preparing it, then derives the specific arguments needed for presenting the results, then presents the results to the user. By the time it gets to the last stage, everything necessary to produce the display should either already be looked up, or else is wrapped in a generator that will fetch them as needed a page at a time, usually, but not always, with a pyinat [Paginator](https://pyinaturalist.readthedocs.io/en/stable/modules/pyinaturalist.paginator.html#paginator) at the bottom layer.

```py
from dronefly.core.commands import Command
from dronefly.core.query import Query
from dronefly.discord.menus import TaxonWithCounts
from pyinaturalist import Place, User

taxon_command = Command("taxon")
try:
    # parsing and preparation:
    # - more properly handled in a context manager; inline here for illustrative purposes
    query = Query.parse('fungi by me from ns in prj lichens atlantic on today")
    prepared_query = query.prepare() # => PreparedQuery

    # get menu arguments:
    taxon = prepared_query.taxon() # => single Taxon that best matches the query
    taxon_count_type = taxon_command.preferred_count_type(prepared_query) # => Union[Type[Place], Type[User], None]
    if taxon_count_type is None:
        taxon_counts = None
    else:
        taxon_counts = get_taxon_counts(prepared_query, taxon_count_type) # => collection of TaxonCounts for counted type

    # start the menu:
    menu = TaxonWithCounts(taxon=taxon, counts=taxon_counts, taxon_count_type=taxon_count_type)
    menu.start()
except:
    # error handling for malformed query, no matching taxon, no matching place, user, etc.
```

A bit of magic in our current `,taxon` command is split up and factored out in this example into `taxon_command.preferred_count_type()` because the type of entity counted by default can vary per command. Then the `get_taxon_counts()` is provided as a helper method on the prepared query itself, given both the `prepared_query` and the `taxon_count_type` as arguments. For instance, as per our current `,taxon` command behaviour, if the `prepared_query` contains a `Place`, then `taxon_count_type = Place`, otherwise `taxon_count_type = User`, and that determines which of the two filters to apply when fetching the counts via the API.

Finally, since the `TaxonCounts` is agnostic regarding whether it was for places or users, we now pass to it the `taxon`, the `counts`, and the `count_type` so the menu labels the table appropriately and also provides the correct buttons to add user or place depending on which type of counts are in the table. The menu shouldn't need to know any of the prepared query details, but instead, everything it needs should be extracted first and passed in as separate arguments.

**Comment by synrg** [comment]
To make `prepare_query()` usable in bot code, any bot-specific lookaside steps (like fetching the user's lookup tables for place & project) just need to be implemented in a subclass of the `dronefly-core.models.Config` that can be created by the bot command via a `@use_config` decorator in the same way as `@use_client` sets up the client, e.g.
```py
@commands.group(invoke_without_command=True)
@checks.bot_has_permissions(embed_links=True)
@use_client
@use_config
async def life(self, ctx, *, query: Optional[Union[TaxonReplyConverter, str]]):
    …
    async with ctx.typing():
        with self.inat_client.set_ctx(ctx) as client:
            try:
                …
                query_response = await prepare_query(client, self.dronefly_config, query)
                obs_args = query_response.obs_args()
                life_list = await client.observations.life_list(**obs_args)
```
Since this part is identical to how core does the command, this excerpt could further be refactored to be used both in the `dronefly-core` life command used by `dronefly-cli` and the bot.

This reintegration back into bot code should be done before we can close this issue. Other ideas above such as the class rename and any ideas for how to restructure commands can be deferred. New issues can be created for those as needed.

**Comment by synrg** [comment]
With the creation of `ContextConfig`, we can now use `prepare_query()` in the bot code.
